### PR TITLE
In JS, this assignments in constructors are preferred and nullable initializers become any

### LINKED ISSUE #22458: In Javascript, null/undefined initializers should be of type any, even with --strict
In Typescript with --strict, an initializer of `null` or `undefined` has its respective type. Without a contextual type, the inferred type is just `null` or `undefined`. This is never desired, but it's easy to add a type annotation. This is not true in Javascript.

```ts
function F(a = null) {
    a // should still be any; null | undefined is a nonsensical type
    this.b = null
}
var f = new F('hi')
f.b = 12

// from async
var q = {
    drain: null,
    process() {
        if (q.drain) {
            q.drain()
        }
    }
}
```

**Expected behavior:**
The call to `new F('hi')`, the assignment `f.b = 12`, and the call `q.drain()` should all be allowed.

**Actual behavior:**
Assignability errors for the first two, "Can't call null" for the third.

### LINKED ISSUE #22792: In JS, type unioning of this-assignments has undesired behaviour
```ts
function A() {
  this.p = 0
}
A.prototype.m = function(x) {
  this.p = x
}
var a = new A()
a.p // p: any
```

**Expected behavior:**
`a.p: number`

**Actual behavior:**
`a.p: any`

This happens because we treat all assignments to the property as declarations and then union their types. And `any | T | ... === any`

But compare:

```ts
function A() {
  this.ps = []
  this.maybe = undefined
}
A.prototype.init = function(x) {
  this.ps = [1,2,3]
  this.maybe = 'just'
}
var a = new A()
a.ps // ps: number[]
a.maybe // maybe: string | undefined
```

When the initialiser in the constructor has an uninformative type, the unioning can result in a much better type, or at least a less bad type (any). But perhaps the fix here is to handle these uninformative initialisers differently.

### PR DESCRIPTION
**Pull request by sandersn**
In Javascript, the compiler unions the type of all special-property-assignments to determine the type of properties declared this way. For example:

```js
function C () { }
C.prototype.m1 = function() { this.a = 1 }
C.prototype.m2 = function() { this.a = 'hi' }
```

This results in C having an instance property `a: string | number | undefined`. However, when there's a this-assignment in the constructor, it should be treated as the definitive declaration:

```js
function C() {
  this.a = 1
}
C.prototype.m1 = function() {
  this.a = 'hi' // error: 'string' is not assignable to 'number'
}
```

Assignment in the constructor is the most common Javascript pattern for declaring instance variables. This PR therefore implements 2 things:

1. Prefer this-assignments in the constructor to determine the type of special assignment declarations.
2. Initializers of type `null`, `undefined` and `never[]` become `any`, `any` and `any[]` respectively. This applies to all initializers, not just special-assignment initializers.

Specifically, when constructing the type for a special-assignment property, if there are this-assignments in the constructor, the types of those this-assignments are used in exclusion of other special-assignments. If, however, these types consist only of `null | undefined`, then all special-assignments are used to construct the type. For example:

```js
function C() {
  this.ctor = 1
  this.ctor = 'hi'
  this.n = null
  this.empty = [] // no other assignments: this.empty: any[]
  this.useless = undefined
}
function C.prototype.m1() {
  this.ctor = false // error: this.ctor: string | number
  this.n = 1 // ok: this.n: number | null
  this.method = 1
  this.method = 'hi' // ok: this.method: string | number | undefined
  this.useless = null // ok, but this.useless: any
}
```

Note that 
1. Properties whose this-assignments only appear in methods add undefined to the type since they may not be initialised during construction.
2. Properties that are assigned nothing but `null` and `undefined` will become `any` and properties that are assigned nothing but `[]` will become `any[]`.

Fixes #22792 
Fixes #22458

I tested this change on the user code in Javascript and it doesn't add any errors that I noticed. I believe this change will provide strictly better typing since the this-assignment-in-constructor pattern is so common in Javascript.

### CODE REVIEW

**Comment by sandersn** [comment]
@DanielRosenwasser I think you might have opinions about this too.

**Comment by weswigham** [comment]
> Initializers of type null, undefined and never[] become any, any and any[] respectively. This applies to all initializers, not just special-assignment initializers.

Flagged as implicit any errors, I hope, so if that's on people know where any's are leaking into their program from?

**Comment by sandersn** [comment]
@weswigham Not yet! Let me think about it.

**Comment by sandersn** [comment]
@weswigham noImplicitAny seems like a good idea. Here is my thought process so far:

1. (Bad) Simple uses of null and undefined are still inconvenient and require type annotation. (`function f(a = null) { }`)
2. (Good) When working with JS, people seem to turn on strictNullChecks before noImplicitAny (at least based on @OliverJAsh's reported experience). Plus, noImplicitAny implies that you *want* to put in all type annotations, even if it's inconvenient.
3. (Bad) The compiler wasn't really designed for strictNullChecks on but noImplicitAny off. We don't *really* know how well it works.
4. (Good) [The compiler comes with free frogurt!](https://www.youtube.com/watch?v=Krbl911ZPBA)

I'm leaning toward putting in the noImplicitAny errors. That's what I'll do next.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -4297,14 +4300,36 @@ namespace ts {
...
-
-            const type = jsDocType || getUnionType(types, UnionReduction.Subtype);
-            return getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
+            let type = jsDocType;
+            if (!type) {
+                const sourceTypes = some(constructorTypes, t => !!(t.flags & ~(TypeFlags.Nullable | TypeFlags.ContainsWideningType))) ? constructorTypes : types;
+                type = getUnionType(sourceTypes, UnionReduction.Subtype);
+            }
+            const widened = getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
+            if (filterType(widened, t => !!(t.flags & ~TypeFlags.Nullable)) === neverType) {
```
Comment (weswigham) [inline src/compiler/checker.ts]: This isn't going to work for generics extending the "nullable" types is it? Eg

```ts
// @filename: m.d.ts
type IsNullable<T extends undefined | null> = T;
type AsNullable<T> = [T] extends [IsNullable<infer U>] ? U : never;
declare function foo<T>(x: T, y: (p: T) => AsNullable<T>): AsNullable<T>;
// @filename: file.js
/**
* @template {T}
* @param {T} x
*/
function Thing(x) {
  this.member = foo(x, x => undefined);
}
```
? (Without generic bounds in js this was the only way I could think to introduce a bounded generic inside js, so maybe this is less of an issue and more of a statement?)

**Pull request merged**